### PR TITLE
fix: #5193 - decorate user-defined zsh widget

### PR DESCRIPTION
**Pull request by po1o**
### Prerequisites

- [x] I have read and understood the [contributing guide][CONTRIBUTING.md].
- [x] The commit message follows the [conventional commits][cc] guidelines.
- [x] Tests for the changes have been added (for bug fixes / features).
- [x] Docs have been added/updated (for bug fixes / features).

### Description

Posh redefines zle-line-init instead of decorating it. This makes some other zsh plugins fail if they also rely on redefining it. This commit adds a helper function to decorate any user-defined widget and use it to decorate zle-line-init.

The logic in this commit is largely inspired by how the zsh-highlighting plugin does the widget wrapping.

This fixes issue #5193.

I tested manually in the dev container by installing zsh-autocomplete and trying with and without this change but I don't know how I could test this automatically. Let me know if you think of a way.

### CODE REVIEW

**Comment by JanDeDobbeleer** [comment]
@po1o that's amazing. Does this also mean we no longer need to fix the broken key bindings below?

**Comment by po1o** [comment]
Oh yes, I hadn't realized there was this workaround.
I believe it's no longer needed indeed.

I'll test when time permits and if it works I will amend my PR

**Comment by JanDeDobbeleer** [comment]
@all-contributors please add @po1o for code

This is pure magic 🪄! @holopin-bot @po1o wizard

**Comment by lewis-yeung** [comment]
@JanDeDobbeleer After this change, when there is no user-defined `zle-line-init` before OMP, the transient prompt in Zsh will not work.

**Comment by JanDeDobbeleer** [comment]
@aha. So the decorate function should check for that I suppose? I'm no expert in zsh (and that documentation is crap).

File: src/shell/scripts/omp.zsh [hunk 1]
```diff
@@ -143,20 +154,43 @@ function enable_poshtooltips() {
...
+  case ${widgets[$cur_widget]:-""} in
+    # Already decorated: do nothing.
+    user:_posh-decorated-*);;
+
+    user:*)
+      zle -N $prefix-$cur_widget ${widgets[$cur_widget]#*:}
+      eval "_posh-decorated-${(q)prefix}-${(q)cur_widget}() { posh::call_widget ${(q)prefix}-${(q)cur_widget} ${(q)posh_widget} -- \"\$@\" }"
+      zle -N $cur_widget _posh-decorated-$prefix-$cur_widget;;
+
+    # For now, do not decorate if it's not a user:*
```
Comment (JanDeDobbeleer) [inline src/shell/scripts/omp.zsh:187]: @lewis-yeung so this should in fact just add it, right?

**Comment by po1o** [comment]
I'll have a look at the comments tonight. I did not have the time to get back to this yet.

File: src/shell/scripts/omp.zsh (hunk 1 above)
Comment (po1o) [reply src/shell/scripts/omp.zsh:187]: Yes - I'll add it.
Sending PR tonight. I had a zle-line-init left in my config when I tested it. I'll use zsh --no-rcs to test. That's my bad

Comment (po1o) [reply src/shell/scripts/omp.zsh:187]: I need to check with ${+widgets[$cur_widget]} if the original widget exists or not.
I will carve out some time tonight for this + testing the bindkey removal.

Comment (lewis-yeung) [reply src/shell/scripts/omp.zsh:187]: > so this should in fact just add it, right?

@JanDeDobbeleer Yes. And I wonder if we can utilize the decorating logic to resolve #5079. Will try a fix when I'm available.

Comment (lewis-yeung) [reply src/shell/scripts/omp.zsh:187]: > I wonder if we can utilize the decorating logic to resolve #5079.

Seems unrelated. 🤷🏻‍♂️

Comment (JanDeDobbeleer) [reply src/shell/scripts/omp.zsh:187]: Yes, that's about key bindings. More tricky that one I guess (or meeds a different approach).

Comment (lewis-yeung) [reply src/shell/scripts/omp.zsh:187]: > I wonder if we can utilize the decorating logic to resolve #5079.

@JanDeDobbeleer Exciting news: this is practical. I've just tested out a fix and it resolved #5079. I'll double check and make a PR in a few days.

**Pull request merged**